### PR TITLE
Optional Additional Loss to Center Reward Models' Outputs

### PR DESCRIPTION
**Pull request by RylanSchaeffer**
In this issue, I requested an optional additional loss function to center the rewards output by a trained reward model: https://github.com/huggingface/trl/issues/1931

This PR is a sketch of what this might look like.

Please let me know if this seems sensible and what changes, if any, are appropriate.

### CODE REVIEW

File: trl/trainer/reward_config.py [hunk 1]
```diff
@@ -37,3 +37,5 @@ class RewardConfig(TrainingArguments):
     max_length: Optional[int] = None
     """The maximum length of the sequences in the batch. This argument is required if you want to use the default data collator."""
     dataset_num_proc: Optional[int] = None
+    """Coefficient to incentivize the reward model to output mean-zero rewards."""
```
Comment (qgallouedec) [inline trl/trainer/reward_config.py:40]:
```suggestion
    """Coefficient to incentivize the reward model to output mean-zero rewards (proposed by https://huggingface.co/papers/2312.09244, Eq. 2)."""
```

Comment (qgallouedec) [reply trl/trainer/reward_config.py:40]: following https://github.com/huggingface/trl/issues/1931#issuecomment-2294902065, is it correct?

File: trl/trainer/reward_trainer.py [hunk 2]
```diff
@@ -244,6 +244,9 @@ def compute_loss(
         else:
             loss = -nn.functional.logsigmoid(rewards_chosen - rewards_rejected).mean()
 
+        if self.args.center_rewards_coefficient is not None:
+            loss += self.args.center_rewards_coefficient * nn.functional.mse_loss(rewards_chosen, rewards_rejected)
```
Comment (qgallouedec) [inline trl/trainer/reward_trainer.py:248]:
```suggestion
            loss += self.args.center_rewards_coefficient * torch.mean((rewards_chosen, rewards_rejected) ** 2)
```

Comment (qgallouedec) [reply trl/trainer/reward_trainer.py:248]: following https://github.com/huggingface/trl/issues/1931#issuecomment-2294902065, is it correct?

Comment (RylanSchaeffer) [reply trl/trainer/reward_trainer.py]: @qgallouedec your implementation is correct.

File: trl/trainer/reward_config.py [hunk 3]
```diff
@@ -37,3 +37,5 @@ class RewardConfig(TrainingArguments):
     max_length: Optional[int] = None
     """The maximum length of the sequences in the batch. This argument is required if you want to use the default data collator."""
     dataset_num_proc: Optional[int] = None
+    """Coefficient to incentivize the reward model to output mean-zero rewards (proposed by https://huggingface.co/papers/2312.09244, Eq. 2)."""
```
Comment (qgallouedec) [inline trl/trainer/reward_config.py:40]:
```suggestion
    """Coefficient to incentivize the reward model to output mean-zero rewards (proposed by https://huggingface.co/papers/2312.09244, Eq. 2). Recommended value:  `0.01`."""
```

**Comment by qgallouedec** [comment]
I'm currently training a model with this new parameter to see what it looks like.

**Comment by RylanSchaeffer** [comment]
I'm hesitant to set the default. The value you chose is correct but I don't
know whether the change should be applied to all users without them
intentionally enabling it. What do you think?

**Comment by qgallouedec** [comment]
> I'm hesitant to set the default.

I wouldn't set it as default either.

**Comment by qgallouedec** [comment]
This one https://github.com/huggingface/trl/pull/1932#discussion_r1720804156 is about suggesting a value in the doc, not setting it as default.

**Comment by RylanSchaeffer** [comment]
Right. I was more asking you for guidance about what default value we should choose. I think you and I are on the same page.

**Comment by RylanSchaeffer** [comment]
@qgallouedec for my education and for posterity's sake, can you share your experimental results here?

**Comment by qgallouedec** [comment]
Sure!

### Training

Here are the wandb runs

`center_rewards_coefficient = None`: https://wandb.ai/huggingface/trl/runs/u6zob8ml
`center_rewards_coefficient = 0.01`: https://wandb.ai/huggingface/trl/runs/d73qlevz

Screenshot 2024-08-17 at 22 04 39

(green is for `center_rewards_coefficient = 0.01`)

As expected the loss is a bit larger.

### Playing with the trained reward model

`center_rewards_coefficient = None`: https://huggingface.co/qgallouedec/reward_modeling_anthropic_hh
`center_rewards_coefficient = 0.01`: https://huggingface.co/qgallouedec/reward_modeling_anthropic_hh_crc


```python
from datasets import load_dataset
from transformers import AutoModelForSequenceClassification, AutoTokenizer

model_id = "qgallouedec/reward_modeling_anthropic_hh"  # without the coef
# model_id = "qgallouedec/reward_modeling_anthropic_hh_crc"  # with the coef

tokenizer = AutoTokenizer.from_pretrained(model_id)
model = AutoModelForSequenceClassification.from_pretrained(model_id, num_labels=1)

dataset = load_dataset("Anthropic/hh-rlhf", split="test")

examples = dataset[:8]

input_chosen = tokenizer(examples["chosen"], return_tensors="pt", padding=True)
input_rejected = tokenizer(examples["rejected"], return_tensors="pt", padding=True)

output_chosen = model(**input_chosen)
output_rejected = model(**input_rejected)

mean_chosen = output_chosen.logits.mean().item()
mean_rejected = output_rejected.logits.mean().item()
print(mean_chosen, mean_rejected)
```

| `center_rewards_coefficient = ` | Chosen  | Rejected |
| ------------------------------- | ------- | -------- |
| `None`                          | -2.1824 | -3.3083  |
| `0.01`                          | 0.2871  | -0.6140  |

So overall it's looking good!

**Comment by qgallouedec** [comment]
I'll just add a little piece of documentation and we're good to merge!

**Review by qgallouedec — APPROVED** [review]

**Comment by qgallouedec** [comment]
Thanks for your first contribution @RylanSchaeffer!
